server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -486,15 +499,16 @@ private static void parseObjectDynamic(DocumentParserContext context, String cur
                     );
                 }
                 if (dynamicObjectMapper instanceof ObjectMapper) {
```

Comment:
Question: here we don't know how the object mapper was created dynamically. In most cases this comes from the usual dynamic mappings (intermediate objects created automatically), but are there cases in which an object may be the result of applying a dynamic template?

The main purpose of this conditional is that we do want to apply dynamic templates that e.g. create ip fields out of objects from docs, and we know implicitly that these can only come from dynamic template because we have no default dynamic mapping for ip fields (or any field that supports parsing objects natively for that matter). I think though that we are also making the assumption that objects are only the result of the default dynamic mappings, while they could be the result of applying a dynamic template. I am not entirely sure what this would cause: such objects could have a different dynamic or enabled property which we are ignoring. I am thinking we may want to consider throwing exception if we get an object mapper as a result of applying a dynamic template. Ideally, we would apply only the dynamic template here so we don't have to guess what happened. 

Does this make sense?

Reply:
I tested this out and you are correct there could be the case in which an object may be result of applying dynamic templates. 
I do agree on the solution of throwing an exception in this case. 
I'll create a PR with all the suggested improvements. 
Thanks for the review!